examples/morpheusvm/cmd/morpheusvm/main.go

```diff
@@ -43,11 +52,54 @@ func main() {
 }
 
 func runFunc(*cobra.Command, []string) error {
-	if err := ulimit.Set(ulimit.DefaultFDLimit, logging.NoLog{}); err != nil {
+	logFactory := logging.NewFactory(logging.Config{
+		DisplayLevel: logging.Debug,
+	})
+
+	log, err := logFactory.Make("main")
+	if err != nil {
+		return fmt.Errorf("failed to initialize log: %w", err)
+	}
+
+	if err := ulimit.Set(ulimit.DefaultFDLimit, log); err != nil {
 		return fmt.Errorf("%w: failed to set fd limit correctly", err)
 	}
 
-	controller, err := controller.New()
+	txDBIndexer := indexer.NewTxDBIndexer(memdb.New())
+	indexerFactory := indexer.NewSubscriptionFactory(txDBIndexer)
+	indexerAPI := indexer.NewAPIFactory(txDBIndexer, "morpheusvm", "/indexer")
+
+	server, handler := rpc.NewWebSocketServer(
+		log,
+		trace.Noop,
+		lconsts.ActionRegistry,
+		lconsts.AuthRegistry,
+		10_000_000,
+	)
+
+	webSocketFactory := rpc.NewWebSocketServerFactory(handler)
+	txRemovedSubscription := rpc.SubscriptionFuncFactory[vm.TxRemovedEvent]{
+		AcceptF: func(event vm.TxRemovedEvent) error {
+			return server.RemoveTx(event.TxID, event.Err)
+		},
+	}
+
+	blockSubscription := &rpc.SubscriptionFuncFactory[*chain.StatelessBlock]{
+		AcceptF: func(event *chain.StatelessBlock) error {
+			return server.AcceptBlock(event)
+		},
+	}
+
+	controller, err := controller.New(
+		vm.WithBlockSubscriptions[*controller.Controller](indexerFactory, blockSubscription),
+		vm.WithTxRemovedSubscriptions[*controller.Controller](txRemovedSubscription),
+		vm.WithVMAPIs[*controller.Controller](
+			rpc.JSONRPCServerFactory{},
+			indexerAPI,
+			webSocketFactory,
+		),
+		vm.WithControllerAPIs[*controller.Controller](&lrpc.JSONRPCServerFactory{}),
+	)
```

Comment:
As discussed yesterday, worth looking at an optional interface pattern where all of the options can optionally implement any of the interfaces, so that they'll be registered by the VM.

In this case, it's important to include any changes in the release notes and for users to include an interface check to make sure that they catch any changes from upstream instead of silently no longer registering their dependencies as desired if the optional interface signature changed.

Reply:
Replaced with the indexer-defined option as Aaron suggested offline.